browser/ai_chat/ai_chat_tab_helper.cc

```diff
@@ -0,0 +1,300 @@
+    : content::WebContentsObserver(web_contents),
+      content::WebContentsUserData<AIChatTabHelper>(*web_contents) {
+  api_helper_ = std::make_unique<AIChatAPIRequest>(
+      web_contents->GetBrowserContext()
+          ->GetDefaultStoragePartition()
+          ->GetURLLoaderFactoryForBrowserProcess());
+}
+
+AIChatTabHelper::~AIChatTabHelper() = default;
+
+std::string AIChatTabHelper::GetConversationHistoryAsString() {
+  std::string history_text;
+
+  for (ConversationTurn history : chat_history_) {
+    history_text = base::StrCat({history_text,
+                                 history.character_type == CharacterType::HUMAN
+                                     ? ai_chat::kHumanPrompt
+                                     : ai_chat::kAIPrompt,
+                                 history.text, "\n"});
+  }
+
+  return history_text;
+}
+
+const std::vector<ConversationTurn>& AIChatTabHelper::GetConversationHistory() {
+  return chat_history_;
+}
+
+void AIChatTabHelper::AddToConversationHistory(const ConversationTurn& turn) {
+  chat_history_.push_back(turn);
+
+  for (auto& obs : observers_) {
+    obs.OnHistoryUpdate();
+  }
+}
+
+void AIChatTabHelper::SetArticleSummaryString(const std::u16string& text) {
+  article_summary_ = text;
+}
+
+void AIChatTabHelper::AddObserver(Observer* observer) {
+  observers_.AddObserver(observer);
+}
+
+void AIChatTabHelper::RemoveObserver(Observer* observer) {
+  observers_.RemoveObserver(observer);
+}
+
+void AIChatTabHelper::RequestSummary() {
+  if (!article_summary_.empty()) {
+    OnArticleSummaryResult(article_summary_, true);
+    return;
+  }
+
+  ui::AXTreeID tree_id = web_contents()->GetPrimaryMainFrame()->GetAXTreeID();
+  content::RenderFrameHost* rfh =
+      content::RenderFrameHost::FromAXTreeID(tree_id);
+
+  rfh->GetMainFrame()->RequestAXTreeSnapshot(
+      base::BindOnce(&AIChatTabHelper::OnSnapshotFinished,
+                     base::Unretained(this)),
+      ui::AXMode::kWebContents, false, 5000, {});
+}
+
+void AIChatTabHelper::OnSnapshotFinished(const ui::AXTreeUpdate& snapshot) {
+  ui::AXTree tree;
+  if (!tree.Unserialize(snapshot)) {
+    return;
+  }
+
+  // Start AX distillation process
+  // Don't copy the tree, as it can be expensive.
+  DistillViaAlgorithm(tree);
+}
+
+void AIChatTabHelper::DistillViaAlgorithm(const ui::AXTree& tree) {
+  std::vector<const ui::AXNode*> content_root_nodes;
+  std::vector<const ui::AXNode*> content_nodes;
+  GetContentRootNodes(tree.root(), &content_root_nodes);
+
+  for (const ui::AXNode* content_root_node : content_root_nodes) {
+    AddContentNodesToVector(content_root_node, &content_nodes);
+  }
+
+  std::vector<std::u16string> text_node_contents;
+  for (const ui::AXNode* content_node : content_nodes) {
+    AddTextNodesToVector(content_node, &text_node_contents);
+  }
+
+  // TODO(nullhook): The assumption here is that 9300 chars equate to
+  // approximately 2k tokens, which is a rough estimate. A proper tokenizer is
```

Comment:
`browser/ai_chat/ai_chat_api_request.cc` sets `max_token_to_sample` to 200, not 2000. Should that setting and this comment use the same number?

Reply:
No, both are different.

The token count I'm referring to in this comment is for the extracted article text, which was estimated using OAI's [tokenizer](https://platform.openai.com/tokenizer). The reason to shorten this to respect Anthropic's context window limit which is sent on every request so the model can generate contextual output.

Max tokens to sample is the length of the generated answer/sequence.